<a href="https://colab.research.google.com/github/nchebet09/Naomi-s-Week-9-IP/blob/master/Naomi's_Week_9_IP%2C_The_Naive_Bayes_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1. Defining the Question

### a) Specifying the Data Analytic Question

Given the text classification data, create a Naive Bayes classification model to predict whether a text is a spam or not

### b) Defining the Metric for Success

To evaluate the perfomance of our model, we will use accuracy score and confusion matrix

### c) Understanding the context 

### d) Recording the Experimental Design

Approach for the project will be as follows:


1. Reading and initial data exploration

2. Data validation

3. Tidying up the data

4. EDA

5. Implementing the solution

6. Model perfomance optimization

7. Conclusion and challenging the solution

### e) Data Relevance

Yes data was relevant to the problem at hand

## 2. Reading the Data

In [46]:
# importing libraries

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.preprocessing import Normalizer

In [2]:
# Loading the Data from the source i.e. csv
spambase = pd.read_csv('spambase_csv.csv')



## 3. Checking the Data

In [3]:
# Determining the no. of records in our dataset
spambase.shape

# our dataset has 4,601 records and 58 columns

(4601, 58)

In [4]:
# Previewing the top of our dataset
spambase.head()

word_freq_make  word_freq_address  ...  capital_run_length_total  class
0            0.00               0.64  ...                       278      1
1            0.21               0.28  ...                      1028      1
2            0.06               0.00  ...                      2259      1
3            0.00               0.00  ...                       191      1
4            0.00               0.00  ...                       191      1

[5 rows x 58 columns]

In [5]:
# Previewing the bottom of our dataset
spambase.tail()

word_freq_make  word_freq_address  ...  capital_run_length_total  class
4596            0.31                0.0  ...                        88      0
4597            0.00                0.0  ...                        14      0
4598            0.30                0.0  ...                       118      0
4599            0.96                0.0  ...                        78      0
4600            0.00                0.0  ...                        40      0

[5 rows x 58 columns]

In [6]:
# Checking whether each column has an appropriate datatype
spambase.dtypes

word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float64
word_freq_000                 float64
word_freq_money               float64
word_freq_hp                  float64
word_freq_hpl                 float64
word_freq_ge

In [7]:
# checking for unique values in each column
spambase.nunique()

# class is our target variable while the rest are features

word_freq_make                 142
word_freq_address              171
word_freq_all                  214
word_freq_3d                    43
word_freq_our                  255
word_freq_over                 141
word_freq_remove               173
word_freq_internet             170
word_freq_order                144
word_freq_mail                 245
word_freq_receive              113
word_freq_will                 316
word_freq_people               158
word_freq_report               133
word_freq_addresses            118
word_freq_free                 253
word_freq_business             197
word_freq_email                229
word_freq_you                  575
word_freq_credit               148
word_freq_your                 401
word_freq_font                  99
word_freq_000                  164
word_freq_money                143
word_freq_hp                   395
word_freq_hpl                  281
word_freq_george               240
word_freq_650                  200
word_freq_lab       

## 4. External Data Source Validation

### a.Validation

## 5. Tidying the Dataset

In [8]:
spambase.columns

Index(['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d',
       'word_freq_our', 'word_freq_over', 'word_freq_remove',
       'word_freq_internet', 'word_freq_order', 'word_freq_mail',
       'word_freq_receive', 'word_freq_will', 'word_freq_people',
       'word_freq_report', 'word_freq_addresses', 'word_freq_free',
       'word_freq_business', 'word_freq_email', 'word_freq_you',
       'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000',
       'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george',
       'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet',
       'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85',
       'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
       'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting',
       'word_freq_original', 'word_freq_project', 'word_freq_re',
       'word_freq_edu', 'word_freq_table', 'word_freq_conference',


In [9]:
# Checking for Anomalies
spambase.duplicated().value_counts()

False    4210
True      391
dtype: int64

In [10]:
# exploring further on duplicated records
duplicated = spambase[spambase.duplicated()]
duplicated
# given the lack of a unique ID, it's hard to tell if these records are actual duplicates. We won't be deleting these records

word_freq_make  word_freq_address  ...  capital_run_length_total  class
26               0.0                0.0  ...                        82      1
103              0.0                0.0  ...                       292      1
104              0.0                0.0  ...                        66      1
105              0.0                0.0  ...                       292      1
106              0.0                0.0  ...                        66      1
...              ...                ...  ...                       ...    ...
4439             0.0                0.0  ...                       282      0
4441             0.0                0.0  ...                       282      0
4537             0.0                0.0  ...                         6      0
4541             0.0                0.0  ...                         2      0
4550             0.0                0.0  ...                         4      0

[391 rows x 58 columns]

In [11]:
# Identifying the Missing Data
print(spambase.isnull().sum())
# our dataset has no missing values

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

## 6. Exploratory Analysis

### Univariate Analysis

In [12]:
spambase.describe()

word_freq_make  word_freq_address  ...  capital_run_length_total        class
count     4601.000000        4601.000000  ...               4601.000000  4601.000000
mean         0.104553           0.213015  ...                283.289285     0.394045
std          0.305358           1.290575  ...                606.347851     0.488698
min          0.000000           0.000000  ...                  1.000000     0.000000
25%          0.000000           0.000000  ...                 35.000000     0.000000
50%          0.000000           0.000000  ...                 95.000000     0.000000
75%          0.000000           0.000000  ...                266.000000     1.000000
max          4.540000          14.280000  ...              15841.000000     1.000000

[8 rows x 58 columns]

In [13]:
# dist_columns = ['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d',
#        'word_freq_our', 'word_freq_over', 'word_freq_remove',
#        'word_freq_internet', 'word_freq_order', 'word_freq_mail',
#        'word_freq_receive', 'word_freq_will', 'word_freq_people',
#        'word_freq_report', 'word_freq_addresses', 'word_freq_free',
#        'word_freq_business', 'word_freq_email', 'word_freq_you',
#        'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000',
#        'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george',
#        'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet',
#        'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85',
#        'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
#        'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting',
#        'word_freq_original', 'word_freq_project', 'word_freq_re',
#        'word_freq_edu', 'word_freq_table', 'word_freq_conference',
#        'char_freq_%3B', 'char_freq_%28', 'char_freq_%5B', 'char_freq_%21',
#        'char_freq_%24', 'char_freq_%23', 'capital_run_length_average',
#        'capital_run_length_longest', 'capital_run_length_total']

# sns.set_style('darkgrid')
# fig, axes = plt.subplots(len(dist_columns), figsize = (10, 15))
# fig.suptitle('Distributions of numerical columns', y= 1.01, color = 'black', fontsize = 15)

# for ax, data, name in zip(axes.flatten(), spambase, dist_columns):
#   sns.distplot(spambase[name], ax = ax, kde = True, color = 'purple')
# plt.tight_layout()

With too many variables it's hard to tell if the features are normally distributed or not. But since this is a text classification, we will use Multinomial Naive Bayes classifier with the assumption that the features are multinomially distributed

### Bivariate Analysis

In [14]:
# Heatmap
# fig, ax = plt.subplots(figsize=(15,15))
# sns.heatmap(spambase.corr(), annot=True, ax=ax);

There are too many variables hence it's hard to read on the correlations. Let's do a correlation table instead

In [33]:
# correlations
spambase.corr()

word_freq_make  ...     class
word_freq_make                    1.000000  ...  0.126208
word_freq_address                -0.016759  ... -0.030224
word_freq_all                     0.065627  ...  0.196988
word_freq_3d                      0.013273  ...  0.057371
word_freq_our                     0.023119  ...  0.241920
word_freq_over                    0.059674  ...  0.232604
word_freq_remove                  0.007669  ...  0.332117
word_freq_internet               -0.003950  ...  0.206808
word_freq_order                   0.106263  ...  0.231551
word_freq_mail                    0.041198  ...  0.138962
word_freq_receive                 0.188459  ...  0.234529
word_freq_will                    0.105801  ...  0.007741
word_freq_people                  0.066438  ...  0.132927
word_freq_report                  0.036780  ...  0.060027
word_freq_addresses               0.028439  ...  0.195902
word_freq_free                    0.059386  ...  0.263215
word_freq_business                0.081928  ...  0.263204
word_freq_email                   0.053324  ...  0.204208
word_freq_you                     0.128243  ...  0.273651
word_freq_credit                  0.021295  ...  0.189761
word_freq_your                    0.197049  ...  0.383234
word_freq_font                   -0.024349  ...  0.091860
word_freq_000                     0.134072  ...  0.334787
word_freq_money                   0.188155  ...  0.216111
word_freq_hp                     -0.072504  ... -0.256723
word_freq_hpl                    -0.061686  ... -0.232968
word_freq_george                 -0.066424  ... -0.183404
word_freq_650                    -0.048680  ... -0.158800
word_freq_lab                    -0.041251  ... -0.133523
word_freq_labs                   -0.052799  ... -0.171095
word_freq_telnet                 -0.039066  ... -0.126912
word_freq_857                    -0.032058  ... -0.114214
word_freq_data                   -0.041014  ... -0.119931
word_freq_415                    -0.027690  ... -0.112754
word_freq_85                     -0.044954  ... -0.149225
word_freq_technology             -0.054673  ... -0.136134
word_freq_1999                   -0.057312  ... -0.178045
word_freq_parts                  -0.007960  ... -0.031035
word_freq_pm                     -0.011134  ... -0.122831
word_freq_direct                 -0.036095  ... -0.064801
word_freq_cs                     -0.009703  ... -0.097375
word_freq_meeting                -0.026070  ... -0.136615
word_freq_original               -0.024292  ... -0.135664
word_freq_project                -0.022116  ... -0.094594
word_freq_re                     -0.037105  ... -0.140408
word_freq_edu                    -0.034056  ... -0.146138
word_freq_table                  -0.000953  ... -0.044679
word_freq_conference             -0.017755  ... -0.084020
char_freq_%3B                    -0.026505  ... -0.059630
char_freq_%28                    -0.021196  ... -0.089672
char_freq_%5B                    -0.033301  ... -0.064709
char_freq_%21                     0.058292  ...  0.241888
char_freq_%24                     0.117419  ...  0.323629
char_freq_%23                    -0.008844  ...  0.065067
capital_run_length_average        0.044491  ...  0.109999
capital_run_length_longest        0.061382  ...  0.216097
capital_run_length_total          0.089165  ...  0.249164
class                             0.126208  ...  1.000000

[58 rows x 58 columns]

## 7. Implementing the Solution

### Multinomial Naive Bayes Classification

#### With a 20% test size

In [29]:
# splitting data into features and target variables

X = spambase.iloc[:,:-1].values
y = spambase['class'].values

# splitting the dataset into training and testing data

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=0)

# using the multinomial bayes classifier and making predictions

clf = MultinomialNB()
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
# evaluating the model
print(confusion_matrix(y_test, y_pred))
print('Mean Accuracy:',round(np.mean(y_pred==y_test)*100,2),'%')

[[454  84]
 [ 89 294]]
Mean Accuracy: 81.22 %


With a 20% test size the model has a mean prediction accuracy of 81.22%.

#### With a 30% test size

In [30]:
# splitting data into features and target variables

X = spambase.iloc[:,:-1].values
y = spambase['class'].values

# splitting the dataset into training and testing data

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3, random_state=0)

# using the multinomial bayes classifier and making predictions

clf = MultinomialNB()
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
# evaluating the model
print(confusion_matrix(y_test, y_pred))
print('Mean Accuracy:',round(np.mean(y_pred==y_test)*100,2),'%')

[[699 123]
 [140 419]]
Mean Accuracy: 80.96 %


With 30% test size the model's accuracy has reduced slightly to 80.96%

#### With a 40% test size

In [31]:
# splitting data into features and target variables

X = spambase.iloc[:,:-1].values
y = spambase['class'].values

# splitting the dataset into training and testing data

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.4, random_state=0)

# using the multinomial bayes classifier and making predictions

clf = MultinomialNB()
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
# evaluating the model
print(confusion_matrix(y_test, y_pred))
print('Mean Accuracy:',round(np.mean(y_pred==y_test)*100,2),'%')

[[952 145]
 [205 539]]
Mean Accuracy: 80.99 %


With 40% test size the model's accuracy is slighly higher than when using 30% test size, but still lower than 20% test size. Our best model so far is with 20% test size

### Improving the models perfomance through feature selection

In [35]:
# Feature selection can be done by dropping highly correlated features
# creating a copy for feature selection
spambase_df = spambase.copy()

# Create correlation matrix
corr_matrix = spambase_df.corr().abs()

# Select upper triangle of correlation matrix
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))

# Find index of feature columns with correlation greater than 0.95
to_drop = [column for column in upper.columns if any(upper[column] > 0.95)]

In [40]:
spambase_df = spambase_df.drop(spambase_df[to_drop], axis=1)

In [41]:
# checking to see if any of the columns were dropped
spambase_df.shape

(4601, 57)

One column that was highly correlated has been dropped. We will now fit the data to MultinomialNB and see if the model's accuracy improves

In [45]:
# splitting data into features and target variables

X = spambase_df.iloc[:,:-1].values
y = spambase_df['class'].values

# splitting the dataset into training and testing data

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=0)

# using the multinomial bayes classifier and making predictions

clf = MultinomialNB()
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
# evaluating the model
print(confusion_matrix(y_test, y_pred))
print('Mean Accuracy:',round(np.mean(y_pred==y_test)*100,2),'%')

[[454  84]
 [ 90 293]]
Mean Accuracy: 81.11 %


Fitting the new data after feature selection the mean accuracy model prediction is at 81.11%, this accuracy is slightly lower than what we had earlier when we used the same test size sample. We can conclude that the first model without feature selection is better as it perfomed better.

### Improving the models perfomance through normalization

In [48]:
# making a copy of df for normalization
spambase_norm = spambase_df.copy()
spambase_norm

word_freq_make  word_freq_address  ...  capital_run_length_total  class
0               0.00               0.64  ...                       278      1
1               0.21               0.28  ...                      1028      1
2               0.06               0.00  ...                      2259      1
3               0.00               0.00  ...                       191      1
4               0.00               0.00  ...                       191      1
...              ...                ...  ...                       ...    ...
4596            0.31               0.00  ...                        88      0
4597            0.00               0.00  ...                        14      0
4598            0.30               0.00  ...                       118      0
4599            0.96               0.00  ...                        78      0
4600            0.00               0.00  ...                        40      0

[4601 rows x 57 columns]

In [50]:
# splitting data into features and target variables
X = spambase_norm.iloc[:,:-1].values
y = spambase_norm['class'].values

# Create normalizer
normalizer = Normalizer(norm='l2')

# Transform feature matrix
normalizer.transform(X)

# splitting the dataset into training and testing data

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=0)

# using the multinomial bayes classifier and making predictions

classifier = MultinomialNB()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)
# evaluating the model
print(confusion_matrix(y_test, y_pred))
print('Mean Accuracy:',round(np.mean(y_pred==y_test)*100,2),'%')

[[454  84]
 [ 90 293]]
Mean Accuracy: 81.11 %


Even after normalization the mean accuracy of our model still remains at 81.11%

## 8. Conclusion and challenging the solution

After creating different kinds of models, we conclude that the best model to use for this classification problem is model 1. Model 1 is where we used 20% of the test size sample. Here we got the highest mean accuracy score of 81.22%. In an attempt to improve the model with feature selection and normalization, the accuracy dropped a bit to 81.11%. These two techniques used did not improve our model's perfomance, and therefore, we recommend trying other techniques such as:

*   Laplace Correction

